### PR TITLE
Batch up and flatten diagnostic updates.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
In response to https://github.com/dotnet/roslyn/pull/22920#issuecomment-346950158

This is a tweak to the diagnostic taggers to improve how they respond to a flood of notifications from the underlying diagnostic services.  There are a few improvements as part of this PR.  

1. We batch up diagnostic notifications over a 50ms window and process them all at once, instead of enqueing and processing each notification on the UI thread individually.
2. Because we're batching, we can do a lot of preprocessing of the notifications to throw out information that has been superceded.  For example, if we get a notification that we're actually removing all the diagnostics for a provider, we can just ignore all previous diagnostic-creates that we haven't processed yet.  Also, any more diagnostic creations we hear about will supercede previous diagnostic creations for that provider.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
tagging @heejaechang @sharwell @jasonmalinowski

File: src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs [hunk 1]
```diff
@@ -207,10 +207,65 @@ private void OnDiagnosticsUpdated(object sender, DiagnosticsUpdatedArgs e)
...
             }
 
+            // We may get a flood of diagnostic notificaitons on the BG.  In order to not overwhelm
+            // the UI thread with work to do, and to avoid excess allocations, we batch up these
+            // notifications, and attempt to process them all at once on the UI thread every 50ms
+            // or so.
+            //
+            // Because we're batching, we can optimize things such that we only store two pieces
+            // of data per provider.  Specifically, if they were removing all diagnostics, and
+            // then the last diagnostics they wanted to create.
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs]: note: it may be possible to optimize things even further.  i.e. just store the last DiagnosticsUpdateArgs for a provider.  However, i wasn't sure if that would end up breaking anything.  i.e. it's not clear to me if there are circumstances where we depend on processing the 'remove' before the 'create'.  So i wanted to make the safest change possible.

File: src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs [hunk 2]
```diff
@@ -207,10 +210,72 @@ private void OnDiagnosticsUpdated(object sender, DiagnosticsUpdatedArgs e)
...
+                    switch (e.Kind)
+                    {
+                    case DiagnosticsUpdatedKind.DiagnosticsRemoved:
+                        // we're being told about diagnostics going away because a document/project
+                        // was removed.  This supercedes all previous removes and creates for this
+                        // id.
+                        batchedUpdates.removeArgs = e;
+
+                        // Clear all all creates for this provider. 
+                        batchedUpdates.createArgs.Clear();
```
Comment (sharwell) [inline src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs]: ❗️ This logic isn't correct: `DiagnosticsRemoved` can be specific to a particular project or document. We need to do two things here:

1. We need to modify data structures as necessary to implement the equivalent of this:

    ```csharp
    batchedUpdates.removeArgs.UnionWith(e);
    ```

2. We need an efficient way to only clear items from `createArgs` that overlap with `e` (and I mean `e` as opposed to `removeArgs`).

File: src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs [hunk 3]
```diff
@@ -207,10 +210,72 @@ private void OnDiagnosticsUpdated(object sender, DiagnosticsUpdatedArgs e)
...
+                        break;
+                    case DiagnosticsUpdatedKind.DiagnosticsCreated:
+                        // We're creating diagnostics. This supercedes all existing creations for
+                        // this id.  However, any existing removal for this id will still happen.
+                        var document = e.Solution.GetDocument(e.DocumentId);
+                        batchedUpdates.createArgs[document] = e;
```
Comment (sharwell) [inline src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs]: ❓ Why not use `DocumentId` as the key here? The use of `Document` means entire solutions could end up rooted in the work queue.

File: src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs]: This is true.  While we remove all diagnostics when we get a DiagnosticsRemoved request:

Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs]: You're right.  While we do remove all diagnostics when we get a DiagnosticsRemoved request:

https://github.com/dotnet/roslyn/blob/1096194309547507cb3f1b8d2fcc3074c8432bab/src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs#L340-L350

We only do this for a removed request that matches our document id.  I need to ensure the latter.

File: src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs]: it's rooted no matter what.  e holds onto a solution.  So Document was more natural here, without changing what is rooted.

**Comment by CyrusNajmabadi** [comment]
> This pull request substitutes the use of IForegroundNotificationService for task scheduling. The service provides an advantage due to time-boxed incremental processing. In terms of the current pull request, this is roughly equivalent to having a Yield() in the inner loop of OnDiagnosticsUpdatedOnForeground.

This is true.  *However* due to the work i've done here that time boxing-incremental-processing is almost certainly unnecessary.  Due to the collapsing of work that i've done, when we wake up we will have at most *two* actual pieces of work to do.  The removal of diagnostics and the creation of diagnostics.  This is because we are attached to a single document, and so the majority of notifications we hear about will simply be thrown away immediately (in a manner of microseconds) when we compare our workspace+docId against the workspace+docId in the event.  

Because we only have two pieces of work to do, doing work incrementally isn't really necessary as no single piece of work should be substantial enough to warrant yielding.

**Comment by CyrusNajmabadi** [comment]
> I'd like to find a way to remove the Task.Delay(50). We want batching to occur when items are building in the queue, but I don't see a particular advantage to inserting a processing delay to "help" this.

Note: the FGNotificationService has a 50ms delay between processing batches of items.  I was preserving that here.

Note that I'd like at least some delay that balances not being noticeable, while still likely catching a huge burst of activity from the diagnostics service.  50ms seems to fit the bill, and was not really a problem in the old system.

File: src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs]: #fixed

**Comment by CyrusNajmabadi** [comment]
@sharwell ready for another pass.

File: src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs]: > It's possible the logic is correct based on information which was not obvious to me at the time of review.

Nope.  You're correct.  I missed something.  Removal notifications for other documents would cause me to remove the diagnostics for our document.

**Comment by CyrusNajmabadi** [comment]
> Because we only have two pieces of work to do,

Ah no.  That's not quite accurate.  We have two pieces of work to do per provider that reported events for this document.  @heejaechang How many different providers are there?  Is there like a provider per analyzer?  (so unbounded in length) Or is it more like just a small handful of providers (i.e. one for the compiler, one for all analyzers, etc.).

Thanks!

**Comment by CyrusNajmabadi** [comment]
At sharwell, i can add the "time-boxed incremental processing" back in if necessary.  Though i'm skeptical it will be necessary.   Note that very little happens when we actually do processing.  All we basically do is go through the list of work (throwing out things unrelated ot our document).  For work items that apply, all that happens is that the underlying AbstractDiagnosticsTaggerProvider.TaggerProvider simply stores the data and then notifies the TagSource that it changed.  The TagSource then in turn does nothing but register a callback to process those tags at some delay in the future (the delay is determined by the type of the diagnostics tag).  

So, basically, we do very little work on the UI thread.

**Comment by CyrusNajmabadi** [comment]
Lol.  @sharwell not "at sharwell" :D

File: src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs [hunk 4]
```diff
@@ -207,10 +209,78 @@ private void OnDiagnosticsUpdated(object sender, DiagnosticsUpdatedArgs e)
...
+
+                    // Check if there's already an in-flight update task.  If so, there's nothing we
+                    // need to do.  The in-flight task will pick up the work we enqueued.  Otherwise,
+                    // create a task to process this work (and anything else that comes in) for 50ms
+                    // from now.
+                    if (_updateTask == null)
+                    {
+                        var token = _owner._listener.BeginAsyncOperation(GetType().Name + "OnDiagnosticsUpdatedOnBackground");
+
+                        _updateTask = Task.Delay(50).ContinueWith(
```
Comment (heejaechang) [inline src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs]: what cancel previous update task if same document changed twice?

File: src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs [hunk 5]
```diff
@@ -207,10 +209,78 @@ private void OnDiagnosticsUpdated(object sender, DiagnosticsUpdatedArgs e)
...
+
+            private readonly object _gate = new object();
+            private ProviderAndDocumentToBatchedUpdates _idToBatchedUpdates = s_providerPool.Allocate();
+            private Task _updateTask = null;
+
             private void OnDiagnosticsUpdatedOnBackground(DiagnosticsUpdatedArgs e)
             {
                 this.AssertIsBackground();
-                RegisterNotification(() => OnDiagnosticsUpdatedOnForeground(e));
+                if (_disposed)
```
Comment (heejaechang) [inline src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs]: shouldn't this under lock?

File: src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs (hunk 4 above)
Comment (heejaechang) [reply src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs]: I see so this is not actually batching, it just update every 50ms.

File: src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs [hunk 6]
```diff
@@ -227,32 +297,91 @@ private void OnDiagnosticsUpdatedOnBackground(DiagnosticsUpdatedArgs e)
...
+                    if (_disposed)
+                    {
+                        return;
+                    }
+
+                    // Do some quick checks to avoid doing any further work for diagnostics  we don't
+                    // care about.
+                    var ourDocument = _subjectBuffer.AsTextContainer().GetOpenDocumentInCurrentContext();
+                    var ourDocumentId = ourDocument?.Id;
+                    if (ourDocumentId != _currentDocumentId)
```
Comment (heejaechang) [inline src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs]: I would prefer doing this in BG by listening to all proper workspace events on buffer/workspace changes notifications. if those events are broken, we should fix those events.

File: src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs]: > what cancel previous update task if same document changed twice?

No need to cancel.  When the task actually executes, it will handle all the notifications that came in between now and then.

File: src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs]: No need for lock here.  At worst, we miss some write of 'true' to disposed.  But then the UI request that executes will immediately bail.

File: src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs (hunk 6 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs]: We could do that as a later change.  It will make things more complex as we'll have to serialize out our BG requests and filter them against the requests to update _currentDocId.  i'm ok doing that as part of a followup PR.

**Comment by CyrusNajmabadi** [comment]
@heejaechang Can you address this part: https://github.com/dotnet/roslyn/pull/23377#issuecomment-347034365

**Comment by heejaechang** [comment]
each analyzer can have 3 different events (1 syntax, 1 semantic, 1 semantic on other files). so in current design, 1 analyzer can have 3 taggers. in reality, except compiler analyzer which do all these 3, most of user analyzer produce only 1 kind of diagnostics (syntax, semantic, or 1 semantic on other files ex, compilation end analyzer) so we can assume 1 tagger per 1 analyzers most of time.

File: src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs (hunk 6 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs]: Note: can you tell me precisely which Workspace/buffer changes I would need to listen to?

**Comment by CyrusNajmabadi** [comment]
Hrmm.. that doesn't really address what i'm looking for.  Let me rephrase.  DiagnosticUpdateArgs has an 'object Id' property in it.  Normally, how many unique different Ids will there be?   A fixed handful?  Or is it unbounded and could be hundreds or more?   i.e. Are there unique Ids per user analyzer in the system?  i.e. some for Add-Using, some for Generate Method, etc. etc.  Or is it more like just a couple of Ids for compiler diagnostics.  Just a couple for *all* analyzers combined, etc. etc.

Thanks!

**Comment by heejaechang** [comment]
well, I like my approach better. since it actually fix workspace events issue and use them as it supposed to.

it uses cancellation token properly.

it still has a bug where it doesn't refresh properly when workspace is changed.

for me , the pull seems only there so that it can do filtering on the UI thread where I think filtering on UI thread is the bug. once we do filtering on BG, there is no reason to go to the UI thread as much at the first place.

File: src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs]: Note: even under a lock we could 'still' miss things, as this field is always racey if accessed from different threads.  The only important thing is that UI calls respect it and don't do work if it is set to true.  the BG just uses it when it can to avoid unnecessary work, but it's ok to miss.

**Comment by CyrusNajmabadi** [comment]
> for me , the pull seems only there so that it can do filtering on the UI thread where I think filtering on UI thread is the bug.

It really shoudln't be.  Filtering should be staggeringly fast.  As mentioned, i can also move it to the BG if i can get info on precisely what events to listen for here.  

The core thing this fixes is insane allocations caused by all the Foreground notifications.  It also properly collapses superceded notifications.  These are the primary things that need fixing.  Adding a couple of microseconds to what happens on the UI thread isn't critical.

File: src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs (hunk 4 above)
Comment (heejaechang) [reply src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs]: yep. I thought you were cancelling previous task and enqueue new one so that it gets delayed again. but this is just putting 50ms buffer on each update.

Comment (heejaechang) [reply src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs]: this still don't cancel update on tagger dispose though.

File: src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs (hunk 6 above)
Comment (heejaechang) [reply src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs]: ContextChange and WorkspaceChanged event.

**Comment by CyrusNajmabadi** [comment]
To me, the critical thing is keeping things simple and clearly delineating lifetimes and data flow properly.  Tagging is super critical and enormously easy to introduce subtle bugs into that are amazingly hard to diagnose and fix.  Keeping things simple, while ensuring we dont' have efficiency problems (like creating 10k+ lambda callbacks in the notification service) is the healthy approach to keep this part of the codebase manageable.  We don't want the tagger infrastructure turning into one of those areas in our code that everyone is afraid to touch because its so complex and hard to understand :)

**Comment by CyrusNajmabadi** [comment]
@heejaechang 

https://github.com/dotnet/roslyn/pull/23377#issuecomment-347363291
^

**Comment by heejaechang** [comment]
"The core thing this fixes is insane allocations caused by all the Foreground notifications." but the cause of that is due to filtering on UI thread.

..

DiagnosticUpdateArgs has an 'object Id' property in it. 1 analyzer can have at most 3 different object Id. that's why I said, in current design, 1 analyzer can have 3 taggers per a file.

**Comment by CyrusNajmabadi** [comment]
> DiagnosticUpdateArgs has an 'object Id' property in it. 1 analyzer can have at most 3 different object Id. that's why I said, in current design, 1 analyzer can have 3 taggers per a file.

That's not answering my question :)  My question is:  around how many unique 'Ids' are there in a normal project?  Is it a handful?  Or is it unbounded?

**Comment by CyrusNajmabadi** [comment]
> but the cause of that is due to filtering on UI thread.

No, that's not true.  I have filtering on the UI thread in this PR without the need for excessive lambda allocations.  So there is no reason that UI thread filtering means excess allocations.

**Comment by heejaechang** [comment]
object id is unique per analyzer, per kind, per document/project, per vsix/nuget analyzer came from.

there will be at least as many object id as number of analyzer, documents, projects in the solution.

**Comment by CyrusNajmabadi** [comment]
> there will be at least as many object id as number of analyzer, documents, projects in the solution.

Ok.  So it's effectively unbounded.  You're telling me that in something like Roslyn, there would be thousands of unique Ids, yes?

**Comment by CyrusNajmabadi** [comment]
> object id is unique per analyzer

I don't know what that means.  Are there a lot of analyzers?  Or are there a fixed number.  For example, do you have one analyzer that handles thigns like Add-Using/Generate-Method, or is it an analyzer per each of these?

**Comment by heejaechang** [comment]
you can remove the excessive lambda allocations by filtering on BG? since it is the tagger that create and enqueue that many lambda when it doesn't need to.

**Comment by CyrusNajmabadi** [comment]
> you can remove the excessive lambda allocations by filtering on BG

And i can remove it on the FG.

**Comment by heejaechang** [comment]
we don't know how many analyzers there will be. it is the diagnostic analyzer we have been making framework for. user can install as many analyzers as they wish. we do not put limit on how many analyzers they can install to VS or solution.

okay, so at most, number of object Id you can get is number of diagnostic analyzers installed (vsix or nuget) * 3 kinds (syntax, semantic, semantic on other files) * number of (documents + projects in the solution) * number of (analyzer vsix * nugets installed)

**Comment by heejaechang** [comment]
sure, you can remove it on FG, but once you do filtering on BG, you don't need to do it on FG since there won't be as many FG requests as before...

so, you don't need to have dictionaries to hold onto batched requests or the delay and etc.

**Comment by heejaechang** [comment]
we wanted to have 1 tagger for 1 analyzers. but current editor tagger doesn't allow dynamic tagger insertion/deletion once view is created. so we made this aggregating tagger concept and created this nested tagger per 1 analyzers I believe. which we can dynamically add and remove, while we give this aggregated tagger to the editor.

I am not 100% sure whether that 1 tagger per analyzer deal with all different events from that analyzer or we have 1 tagger per each different type of analyzer events (aka object id)

**Comment by CyrusNajmabadi** [comment]
I have a followup PR here that moves filtering to the BG as well: https://github.com/dotnet/roslyn/pull/23409

**Comment by CyrusNajmabadi** [comment]
Superseded with https://github.com/dotnet/roslyn/pull/23409

**Comment by CyrusNajmabadi** [comment]
> I am not 100% sure whether that 1 tagger per analyzer deal with all different events from that analyzer or we have 1 tagger per each different type of analyzer events (aka object id)

We have one per Id.  So it's whatever granularity the diagnostics service creates Ids for.  We can't do any better on our end as that's all the information we're given from DiagnosticUpdateArgs

**Comment by CyrusNajmabadi** [comment]
@Pilchie @sharwell This is a much less risky approach to deal with the diagnostic tagging perf issue.  It would probably still be good to smoke test.  It would be safer to take initially, and then could be followed in a later PR by https://github.com/dotnet/roslyn/pull/23409

File: src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs]: That's true.  But worse case that means we just process one more request within 50 ms after tagger dispose. this request will see htat _dispose is set and will immediately bail.  Basically, we have to ask ourselves how important is it to cancel a single request.

**Comment by sharwell** [comment]
@CyrusNajmabadi I'm more confident with #23409, and would prefer to focus the validation on that change.

**Comment by CyrusNajmabadi** [comment]
@sharwell That's fine.  I just wanted to give you guys options depending on your risk aversion.

**Comment by CyrusNajmabadi** [comment]
Closing in favor of: https://github.com/dotnet/roslyn/pull/23448
